fix compilation on macOS when clang<9 & Qt<5.12

tenzap opened the pull request:
1. The code uses `@available` syntax which is new in clang 9
This prevents compilation on older version of macOS that don't
use this version. For example on El Capitan.
This fix will allow to compile on such older systems.


2. The code uses 'QOperatingSystemVersion::MacOSBigSur' which doesn't exist
in some versions (it has been backported to Qt 5.12.10+ & 5.15.1+ only).
On older macos systems like El Capitan the last supported
version of Qt is 5.11
This will fix compilation issue on such older systems and on systems
running with Qt not supporting QOperatingSystemVersion::MacOSBigSur
Compilation error was:
error: no member named 'MacOSBigSur' in 'QOperatingSystemVersion'

3. More details for the other changes are in the commit messages

## Testing strategy
Checked compilation on El Capitan (through macports)

## Type of change
[NOTE]: # ( Please remove all lines which don't apply. )
- ✅ Bug fix (non-breaking change that fixes an issue)

tenzap committed:
fix compilation on Qt not having QOperatingSystemVersion::MacOSBigSur

The code uses 'QOperatingSystemVersion::MacOSBigSur' which doesn't exist
in all Qt versions (it has been backported to Qt 5.12.10+ & 5.15.1+ only).
On older macos systems like El Capitan the last supported
version of Qt is 5.11

This will fix compilation issue on such older systems and on systems
running with Qt not supporting QOperatingSystemVersion::MacOSBigSur

Compilation error was:
error: no member named 'MacOSBigSur' in 'QOperatingSystemVersion'

tenzap committed:
Fix compilation when osx <= 10.9

* AppKitImpl.mm: button property is new in 10.10. It is used for a feature of
  KeePassXC that is only available from 10.17 onwards. So we don't need it when
  compiling on <= 10.9
error: property 'button' not found on object of type 'NSStatusItem *'
        NSString* appearance = [dummy.button.effectiveAppearance.name lowercaseString];
                                      ^

* The code uses @available syntax which is supported by
  AppleClang >= 9 or LLVM >= 5.
  We check __clang_major__ to allow compilation on older versions
  of macOS that don't have a recent clang. For example on El Capitan.

tenzap committed:
Fix compilation when osx <= 10.8

* AppKitImpl.mm: AXIsProcessTrustedWithOptions exists from 10.9 onwards
error: use of undeclared identifier 'kAXTrustedCheckOptionPrompt'
error: use of undeclared identifier 'AXIsProcessTrustedWithOptions'

tenzap committed:
Fix compilation when osx <= 10.7

* MacUtils.cpp: CoreGraphics exists from 10.8 onwards only, capslock detection
  feature would have to be implemented on OSX <= 10.7

* AppKitImpl.mm: CGDisplayStreamRef exists from 10.8 onwards only. It is used for a
  feature of KeePassXC that is only available from 10.15 onwards. So we don't need it
  when compiling on <= 10.7
error: unknown type name 'CGDisplayStreamRef'

* AppKitImpl.mm: Syntax is not understood by 10.7, update it to be understandable
  by <= 10.7
error: expected method to read dictionary element not found on object of type 'NSDictionary *'
    NSRunningApplication* app = userInfo[NSWorkspaceApplicationKey];
                                ^

* The code uses @available syntax which is supported by
  AppleClang >= 9 or LLVM >= 5.
  We check __clang_major__ to allow compilation on older versions
  of macOS that don't have a recent clang.

tenzap commented:
Are the failing tests in CI normal ? I don't see where they could come from.

droidmonkey commented:
Yah we have a lot of problems with the gui tests randomly failing.

tenzap committed:
Fix compilation error on OS X 10.11

src/core/Alloc.cpp:44:10: error: no type named 'free' in namespace 'std'
    std::free(ptr);
    ~~~~~^

This is a regression, since it was fixed in [1]

Per [2], std::free() needs #include <cstdlib>. That file is included
indirectly on newer systems.

[1] https://github.com/keepassxreboot/keepassxc/commit/7c6c027d33b06eb706f93e3d178a2305d7bcfd56
[2] https://en.cppreference.com/w/cpp/memory/c/free

droidmonkey commented on src/gui/osutils/macutils/MacUtils.cpp:
```diff
@@ -138,7 +141,11 @@ void MacUtils::setLaunchAtStartup(bool enable)
 
 bool MacUtils::isCapslockEnabled()
 {
+#if defined MAC_OS_X_VERSION_10_8
```
Is this defined for all future versions as well?

tenzap replied:
Although I didn't find any clear documentation on these macros, it looks like they are present in newer versions of the SDK. I don't have a recent mac, but for older SDKs, each SDK version holds the  MAC_OS_X_VERSION_* of the previous versions. 

So they are apparently defined in future versions.
 
For instance:
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk.10.7/usr/include/AvailabilityMacros.h
```
/*
 * Set up standard Mac OS X versions
 */
#define MAC_OS_X_VERSION_10_0 1000
#define MAC_OS_X_VERSION_10_1 1010
#define MAC_OS_X_VERSION_10_2 1020
#define MAC_OS_X_VERSION_10_3 1030
#define MAC_OS_X_VERSION_10_4 1040
#define MAC_OS_X_VERSION_10_5 1050
#define MAC_OS_X_VERSION_10_6 1060
#define MAC_OS_X_VERSION_10_7 1070
```
/Library/Developer/CommandLineTools/SDKs/MacOSX10.12.sdk/usr/include/AvailabilityMacros.h
```
/*
 * Set up standard Mac OS X versions
 */
#define MAC_OS_X_VERSION_10_0         1000
#define MAC_OS_X_VERSION_10_1         1010
#define MAC_OS_X_VERSION_10_2         1020
#define MAC_OS_X_VERSION_10_3         1030
#define MAC_OS_X_VERSION_10_4         1040
#define MAC_OS_X_VERSION_10_5         1050
#define MAC_OS_X_VERSION_10_6         1060
#define MAC_OS_X_VERSION_10_7         1070
#define MAC_OS_X_VERSION_10_8         1080
#define MAC_OS_X_VERSION_10_9         1090
#define MAC_OS_X_VERSION_10_10      101000
#define MAC_OS_X_VERSION_10_10_2    101002
#define MAC_OS_X_VERSION_10_10_3    101003
#define MAC_OS_X_VERSION_10_11      101100
#define MAC_OS_X_VERSION_10_11_2    101102
#define MAC_OS_X_VERSION_10_11_3    101103
#define MAC_OS_X_VERSION_10_11_4    101104
#define MAC_OS_X_VERSION_10_12      101200
#define MAC_OS_X_VERSION_10_12_1    101201
#define MAC_OS_X_VERSION_10_12_2    101202
```

Other ways to achive the same are described in Availability.h, AvailabilityMacros.h, on this [wiki](https://wiki.herzbube.ch/index.php/Mac_OS_X_Programming#Availability_Macros), in [TechNote 2064](https://developer.apple.com/library/archive/technotes/tn2064/_index.html#//apple_ref/doc/uid/DTS10003092-CH1-SECTION3), and this [archived SDK Compatibility Guide](https://developer.apple.com/library/archive/documentation/DeveloperTools/Conceptual/cross_development/Using/using.html#//apple_ref/doc/uid/20002000-1114741-CJADDEIB).

tenzap replied:
Actually it seems safer to use `MAC_OS_X_VERSION_MIN_REQUIRED >= ...` instead. So I changed for it.

tenzap commented:
Since KeePassXC builds fine on MacPorts from OSX v10.7 to macOS v11 with these fixes (2.6.6 & develop branch), I may not need to further amend the PR on my side. Any comments are still welcome of course :)

Do you have an idea when this could be merged ?

(I add this comment again because it has "pending" status in the review and don't know how to get out of this status:
> Actually it seems safer to use `MAC_OS_X_VERSION_MIN_REQUIRED >= ...` instead of `MAC_OS_X_VERSION_*`. So I changed for it.)

tenzap committed:
fix compilation when macos SDK <= 10.14

These methods are only available from macOS 10.15
  - kSecAccessControlWatch
  - LAPolicy.deviceOwnerAuthenticationWithBiometricsOrWatch

The code uses @available syntax which is supported by
AppleClang >= 9 or LLVM >= 5.
We check __clang_major__ to allow compilation on older versions
of macOS that don't have a recent clang.

droidmonkey commented:
I am conflicted about all of these changes. OSX (10.11 and earlier) has been end of support/life/etc for 3 years and more. You are operating a completely vulnerable system. I am not in favor of carrying technical debt, by way of build exceptions, operating systems that are completely unsupported. 

To me this is tantamount to supporting windows 7 and earlier.

tenzap commented:
I understand your point.

On my side, please note that the background is that I'm updating the Portfile to build KeePassXC on MacPorts.

Although osx 10.11 and other later/earlier versions aren't supported anymore by Apple, it doesn't mean it is not used anymore. Unfortunately some older macs which are still totally usable nowadays can't be updated to newer osx versions.

These changes allow to run KeePassXC on such older macs. Why not allowing compilation on them?

Of course we could keep maintaining these not so complex patches only in macports, but I think it would be better if they were merged upstream. That may benefit someone else too.

droidmonkey commented:
I'll mull this over, I don't like the layered checks: clang + macos define + @available

tenzap commented:
BTW commit 8d842aa is for 10.13 which you officially support according to https://github.com/keepassxreboot/keepassxc/blob/develop/release-tool#L47
`MACOSX_DEPLOYMENT_TARGET=10.13`

tenzap commented:
> I don't like the layered checks: clang + macos define + @available

You probably know this already but I prefer to summarize the changes for reference.

From what I read & understood:
- _@available syntax_ is only understood by [Xcode/AppleClang >= 9](https://developer.apple.com/library/archive/releasenotes/AppKit/RN-AppKit/index.html) or [LLVM >= 5.0](https://clang.llvm.org/docs/LanguageExtensions.html#objective-c-available). Hence the clang define.
- _@available_ is a runtime check only.
- _macos define_ is a compile-time check. Therefore it is needed if we don't want compilation to fail for "unknown/undefined method or whatever" when building with a SDK that doesn't provide that method. For example if we use a method or whatever of 10.15 (like LAPolicyDeviceOwnerAuthenticationWithBiometricsOrWatch) with a SDK for 10.14 then it will fail. Hence the use of the _macos define_ since _@available syntax_ has no effect in this case.

There may be other ways to do these checks, but I didn't find anything better.

phoerious commented:
I don't see why you would need both `@available` and the ifdef guards. There are only three cases:

- new macOS: has new Clang and all needed functions, nothing needs to be done
- old macOS: has somewhat recent Clang and all needed function declarations, but functionality is missing, so `@available` suffices, but ifdef guard would work as well
- ancient macOS: has neither, ifdef guard required, `@available` is redundant

tenzap commented:
It seems `@available` is not sufficient for example on 10.13. On osx10.13 with sdk set to 10.13 & xcode 9, the compiler complains that `kSecAccessControlWatch` and `LAPolicy.deviceOwnerAuthenticationWithBiometricsOrWatch` are missing and stops (even though `@available` is used).

With older versions, down to 10.7, there is the same problem. As soon as the compiler encounters a name that is not known for the target SDK it fails. Maybe there is another way to fix this problem, but that's the only one I found.

Another issue is that if clang is too old (which is the case on older systems) `@available` is not understood and compiler fails.

The objective here is to be able to compile KeePassXC from osx v10.7 up to the latest macos versions (see https://github.com/keepassxreboot/keepassxc/pull/7117#issuecomment-975473272). On MacPorts, with these patches applied it works. Without them compilation fails with the various errors that are reported in each individual commit message.

droidmonkey commented on src/touchid/TouchID.mm:
```diff
@@ -95,18 +95,24 @@ inline QString hash(const QString& value)
     // prepare adding secure entry to the macOS KeyChain
     CFErrorRef error = NULL;
     SecAccessControlRef sacObject;
+#if __clang_major__ >= 9
     if (@available(macOS 10.15, *)) {
+#if MAC_OS_X_VERSION_MIN_REQUIRED >= 101500
```
Can you combine these as well like the others? I like that presentation better and the effect is the same at the end of the day.

tenzap replied:
Done.
